modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1929,49 +1885,50 @@ void GDScriptAnalyzer::reduce_array(GDScriptParser::ArrayNode *p_array) {
-void GDScriptAnalyzer::update_array_literal_element_type(const GDScriptParser::DataType &p_base_type, GDScriptParser::ArrayNode *p_array_literal) {
-	GDScriptParser::DataType array_type = p_array_literal->get_datatype();
-	if (p_array_literal->elements.size() == 0) {
-		// Empty array literal, just make the same type as the storage.
-		array_type.set_container_element_type(p_base_type.get_container_element_type());
-	} else {
-		// Check if elements match.
-		bool all_same_type = true;
-		bool all_have_type = true;
-
-		GDScriptParser::DataType element_type;
-		for (int i = 0; i < p_array_literal->elements.size(); i++) {
-			if (i == 0) {
-				element_type = p_array_literal->elements[0]->get_datatype();
-			} else {
-				GDScriptParser::DataType this_element_type = p_array_literal->elements[i]->get_datatype();
-				if (this_element_type.has_no_type()) {
-					all_same_type = false;
-					all_have_type = false;
-					break;
-				} else if (element_type != this_element_type) {
-					if (!is_type_compatible(element_type, this_element_type, false)) {
-						if (is_type_compatible(this_element_type, element_type, false)) {
-							// This element is a super-type to the previous type, so we use the super-type.
-							element_type = this_element_type;
-						} else {
-							// It's incompatible.
-							all_same_type = false;
-							break;
-						}
-					}
-				}
-			}
+void GDScriptAnalyzer::update_array_literal_element_type(GDScriptParser::ArrayNode *p_array, const GDScriptParser::DataType *p_base_type) {
+	GDScriptParser::DataType array_type = p_array->get_datatype();
+
+	GDScriptParser::DataType base_element_type;
+	bool has_base_element_type = p_base_type != nullptr && p_base_type->has_container_element_type();
+	if (has_base_element_type) {
+		base_element_type = p_base_type->get_container_element_type();
+		base_element_type.is_constant = false;
+	}
+
+	if (p_array->elements.size() == 0) {
+		if (has_base_element_type) {
+			array_type.set_container_element_type(base_element_type);
+			p_array->set_datatype(array_type);
 		}
-		if (all_same_type) {
-			element_type.is_constant = false;
-			array_type.set_container_element_type(element_type);
-		} else if (all_have_type) {
-			push_error(vformat(R"(Variant array is not compatible with an array of type "%s".)", p_base_type.get_container_element_type().to_string()), p_array_literal);
+		return;
+	}
+
+	GDScriptParser::DataType element_type = p_array->elements[0]->get_datatype();
+	if (element_type.has_no_type() || element_type.is_variant()) {
+		return;
+	}
+	for (int i = 1; i < p_array->elements.size(); i++) {
+		GDScriptParser::DataType this_element_type = p_array->elements[i]->get_datatype();
+		if (this_element_type.has_no_type() || this_element_type.is_variant()) {
+			return;
 		}
+		if (is_type_compatible(element_type, this_element_type, false)) {
+			continue;
+		}
+		if (is_type_compatible(this_element_type, element_type, false)) {
+			element_type = this_element_type;
```

Comment:
Why is this needed? You're always doing a reverse check anyway.

Also, it seems simpler to do:
```
if (!is_type_compatible(element_type, this_element_type, false) && !is_type_compatible(this_element_type, element_type, false)) {
  return;
}
```

Reply:
If you have array `[B, A]` then the inferred type is `Array[A]` even though `A` cannot be assigned to `B` (assuming that `B` extends `A`). [Here are those lines](https://github.com/godotengine/godot/blob/6318320c04f348eee1df7a5dae3bad540ef1fdcc/modules/gdscript/gdscript_analyzer.cpp#L1955-L1956) in original code.

Reply:
Ah, I see. So this effectively goes up the class hierarchy as far up as it can? That makes sense to me then, and the code as-is looks good :)

Reply:
Yes, but actually there is a small problem - it is not as smart, so if you have array `[B, C]` where `B` and `C` are separate descendants of `A`, then it will fail, will not infer that it is `Array[A]` unfortunately.

Reply:
Sounds like a fun PR for another day though! :)